### PR TITLE
BUG: df[col] = arr should not overwrite data in df[col]

### LINKED ISSUE #33457: REGR: setting column with setitem should not modify existing array inplace
So consider this example of a small dataframe with a nullable integer column:

```
def recreate_df():
    return pd.DataFrame({'int': [1, 2, 3], 'int2': [3, 4, 5],
                         'float': [.1, .2, .3],
                         'EA': pd.array([1, 2, None], dtype="Int64")
                        })
```

Assigning a new column with `__setitem__` (`df[col] = ...`) normally does not even preserve the dtype:

```
In [2]: df = recreate_df() 
   ...: df['EA'] = np.array([1., 2., 3.]) 
   ...: df['EA'].dtype 
Out[2]: dtype('float64')

In [3]: df = recreate_df() 
   ...: df['EA'] = np.array([1, 2, 3]) 
   ...: df['EA'].dtype
Out[3]: dtype('int64')
```

When assigning a new nullable integer array, it of course keeps the dtype of the assigned values:

```
In [4]: df = recreate_df() 
   ...: df['EA'] = pd.array([1, 2, 3], dtype="Int64") 
   ...: df['EA'].dtype 
Out[4]: Int64Dtype()
```

However, in this case you now also have the tricky side-effect of being in place:

```
In [5]: df = recreate_df() 
   ...: original_arr = df.EA.array 
   ...: df['EA'] = pd.array([1, 2, 3], dtype="Int64") 
   ...: original_arr is df.EA.array  
Out[5]: True
```

I don't think this behaviour should depend on the values being set, and setitem should always replace the array of the ExtensionBlock. 

Because with the above way, you can unexpectedly alter the data with which you created the dataframe. See also a different example using Categorical of this at the original PR that introduced this: https://github.com/pandas-dev/pandas/pull/32831#issuecomment-611994512

cc @jbrockmendel

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #33457
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

cc @jorisvandenbossche this still fails 7 tests locally and there's one more (commented in-line) test that looks fishy.  Extra eyeballs would be welcome.

xref  #35271, #35266

### CODE REVIEW

**Comment by TomAugspurger** [comment]
What's the summary of the behavior change from 1.0.5? `DataFrame.__setitem__[array]` will not mutate the existing array inplace? What dtypes does this affect?

**Comment by TomAugspurger** [comment]
@jbrockmendel I attempted a whatsnew in e60023725693586e8b9b47b9d8eaa14e09ea0ca0, if you could take a look.

**Comment by TomAugspurger** [comment]
Going through the failing tests

* test_fancy_getitem_slice_mixed: OK
* TestDataFrameIndexing.test_iloc_row: OK
* TestDataFrameIndexing.test_iloc_col: OK
* TestiLoc2.test_identity_slice_returns_new_object: ... Probably OK
* TestiLoc2.test_iloc_setitem_categorical_updates_inplace: Probably OK
* TestMerge.test_merge_nocopy: Probably OK

Looking into  `test_apply_function_with_indexing` some more now.

**Comment by TomAugspurger** [comment]
One question on desired behavior: `df.loc[:, "A"] = value` *should* mutate the array inplace, right? On this branch, that is lost:

```python
In [2]: df = pd.DataFrame({"A": [1, 2, 3]})

In [3]: df2 = df.iloc[:]

In [4]: df._mgr.blocks[0].values is df2._mgr.blocks[0].values
Out[4]: True

In [5]: df.loc[:, "A"] = 0

In [6]: df2
Out[6]:
   A
0  1
1  2
2  3
```

**Comment by jbrockmendel** [comment]
> One question on desired behavior: df.loc[:, "A"] = value should mutate the array inplace, right? On this branch, that is lost:

Agreed.  I think thats the behavior with the FIXME comment in test_block_internals.

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -351,6 +351,83 @@ Notable bug fixes
...
+
+.. code-block:: python
+
+   df = pd.DataFrame({"A": [1, 2, 3]})
+   df2 = df[['A']]
+   df['A'] = np.array([0, 0, 0])
+
+Previously, whether ``df2`` was mutated depending on the dtype of the array being assigned to. Now, a
+new array is consistently assigned, so ``df2`` is not mutated.
+
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.1.0.rst:430]: Looks good.  Maybe a sentence about how to get the mutating behavior with `df.loc[:, col] = foo`?

**Comment by jbrockmendel** [comment]
> test_fancy_getitem_slice_mixed: OK

The relevant part of this test reads:

```
        sliced = float_frame.iloc[:, -3:]

        msg = r"\nA value is trying to be set on a copy of a slice from a DataFrame"
        with pytest.raises(com.SettingWithCopyError, match=msg):
            sliced["C"] = 4.0

        assert (float_frame["C"] == 4).all()
```

I think with the new behavior, the last assertion is now incorrect, as we expect the setitem to create a new array.  My question is: do we still need the SettingWithCopyError?

**Comment by TomAugspurger** [comment]
Sorry, but "OK" I meant OK with updating the tests for the new behavior.

>  My question is: do we still need the SettingWithCopyError?

I was wondering this as well. If the warning is about whether or not `float_frame` is updated then it seems like it can be removed. But perhaps the test should be updated to use `sliced.loc[:, "C"] = 4.0`, in which case the warning is still valid?

**Comment by jbrockmendel** [comment]
Down to two failing tests:

```
FAILED pandas/tests/groupby/test_apply_mutate.py::test_apply_function_with_indexing - AssertionError: Series are different
FAILED pandas/tests/reshape/merge/test_merge.py::TestMerge::test_merge_nocopy - AssertionError: assert False
```

`test_apply_function_with_indexing` i've tracked down into `libreduction.apply_frame_axis0`

**Comment by simonjayhawkins** [comment]
@jbrockmendel 1.2 whatsnew now merged. also you can add back the release note that was deleted in #35271?

**Comment by jbrockmendel** [comment]
@TomAugspurger I've got this down to one failing test `test_apply_function_with_indexing`, could use some extra eyeballs.

That test reads:

```
df = pd.DataFrame(
        {"col1": ["A", "A", "A", "B", "B", "B"], "col2": [1, 2, 3, 4, 5, 6]}
)

def fn(x):
        x.col2[x.index[-1]] = 0
        return x.col2

result = df.groupby(["col1"], as_index=False).apply(fn)
expected = pd.Series(
        [1, 2, 0, 4, 5, 0],
        index=pd.MultiIndex.from_tuples(
            [(0, 0), (0, 1), (0, 2), (1, 3), (1, 4), (1, 5)]
        ),
        name="col2",
)
tm.assert_series_equal(result, expected)
```

ATM instead of `expected` I'm getting for `result`:

```
>>> result
0  0    1
   1    2
   2    0
1  3    1
   4    2
   5    0
Name: col2, dtype: int64

```

I can get this to pass if instead of `x.col2[x.index[-1]] = 0` I write `x.loc[x.index[-1], "col2"] = 0`

Alternatively, I get the expected result if I disable `fast_apply` in `groupby.ops`.

**Comment by TomAugspurger** [comment]
Haven't looked closely, but I'm probably OK with changing the test to use
`x.loc[x.index[-1], "col2"] = 0` instead. That's the recommended way to do
things anway.

**Comment by jbrockmendel** [comment]
> I'm probably OK with changing the test [...] That's the recommended way to do things anway.

Makes sense.  But would that be "breaking" the behavior for the non-recommended usage?

**Comment by TomAugspurger** [comment]
Indeed, though it's unclear whether that usage was "officially" supported. IIUC, the `x2.col[x.index[-1]] = 0` might not work for heterogenous dataframes?

**Comment by jorisvandenbossche** [comment]
I would be hesitant to break/change the existing test. With all view/copy complexities, the fact that getting one column as a series is a view on the data is one of the few certainties? 
(and thus editing that series in place should update the original dataframe?)

**Comment by jbrockmendel** [comment]
> might not work for heterogenous dataframes?

I've cooked up an example where it doesn't work for non-unique columns:

```
df = pd.DataFrame(np.random.randn(3, 3), columns=["A", "A", "B"])
ser = df.B
ser[ser.index[-1]] = 0  # <-- produces `SettingWithCopyWarning`, does not modify `df`
```

> the fact that getting one column as a series is a view on the data is one of the few certainties?

The example above shows that this is not the case.  That said, I agree it _should_ be.

> Do you understand the reason why this case is now broken?

I've tracked it down into libreduction.apply_frame_axis0 and am stuck there

**Comment by jbrockmendel** [comment]
OK, so what I think is happening in `libreduction.apply_frame_axis0` is the following:

Instead of creating a bunch of DataFrames for the different groups, it creates one DataFrame, grabs the ndarrays in each block, and for each group moves the pointer so the same array points to the data for that group.

That trick is only valid if the blocks in this DataFrame remain unchanged throughout this iteration.  I think this can be re-pinned at each stage of the loop (albeit at a perf cost)

**Comment by TomAugspurger** [comment]
Thanks for working on this.

> if the blocks in this DataFrame remain unchanged throughout this iteration. I think this can be re-pinned at each stage of the loop (albeit at a perf cost)

Just thinking out loud, can we store property on the Block that somehow tracks if the block has been mutated? Then the for loop can be something like

```python
mutated = False
for block in blocks:
    func(block)
    if block.mutated:
        # do stuff
```

I suppose that would only work for things going through `Block.set`, but maybe that's enough? And feel free to ignore this if it doesn't make any sense in this context.

**Comment by jbrockmendel** [comment]
> Just thinking out loud, can we store property on the Block that somehow tracks if the block has been mutated? Then the for loop can be something like

This is similar to what the commit-before-last (pushed about an hour ago) did

**Comment by jbrockmendel** [comment]
@jorisvandenbossche are there unaddressed issues here?

**Comment by jorisvandenbossche** [comment]
Yes, see the discussion in #33457, I am not convinced that that are changes we should be making in 1.2

File: pandas/tests/frame/test_block_internals.py [hunk 2]
```diff
@@ -644,3 +646,27 @@ def test_to_dict_of_blocks_item_cache():
...
+        {
+            "a": pd.Series([np.nan, 2.0, 3.0, 1.0]).astype("category"),
+            "b": pd.Series(["A", "A", "B", "C"]).astype("category"),
+            "c": pd.Series(["D", "E", "E", np.nan]).astype("category"),
+        }
+    )
+    cats = df["a"].cat.categories.tolist()
+    # append new category
+    cats.append(0.0)
+    df["a"] = df["a"].astype(pd.CategoricalDtype(categories=cats, ordered=False))
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_block_internals.py:663]: Can you use the simpler reproducer from https://github.com/pandas-dev/pandas/issues/35731#issuecomment-674409336 ? (that's sufficient for having the "Series as block values" bug)

**Review by jreback — CHANGES_REQUESTED** [review]
still in discussion

**Comment by jbrockmendel** [comment]
this is a blocker for being able to (accurately) cache properties on Blocks

File: doc/source/whatsnew/v1.2.0.rst [hunk 3]
```diff
@@ -192,6 +192,90 @@ Other enhancements
...
+   >>> df['a'] = 0
+   >>> a  # not mutated
+   <IntegerArray>
+   [1, 2, 3]
+   Length: 3, dtype: Int64
+
+
+*New Behavior*
+
+In pandas 1.1.0, ``DataFrame.__setitem__`` consistently sets on a new array rather than
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.2.0.rst:246]: This needs changing too.

File: pandas/tests/groupby/test_apply.py [hunk 4]
```diff
@@ -985,27 +985,6 @@ def test_apply_function_index_return(function):
     tm.assert_series_equal(result, expected)
 
 
-def test_apply_function_with_indexing():
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_apply.py:988]: this test is duplicated identically in test_apply_mutate

File: doc/source/whatsnew/v1.2.0.rst [hunk 5]
```diff
@@ -341,6 +341,90 @@ instead of casting to NumPy which may have different semantics (:issue:`36076`,
...
+
+
+*New Behavior*
+
+In pandas 1.2.0, ``DataFrame.__setitem__`` consistently sets on a new array rather than
+mutating the existing array inplace.
+
+.. ipython:: python
+
+For NumPy's int64 dtype
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:400]: move before the title (L399)

File: pandas/tests/frame/methods/test_rename.py [hunk 6]
```diff
@@ -164,7 +164,7 @@ def test_rename_multiindex(self):
 
     def test_rename_nocopy(self, float_frame):
```
Comment (jreback) [inline pandas/tests/frame/methods/test_rename.py:165]: can you parameterize this both with and w/o copy

File: pandas/tests/internals/test_internals.py [hunk 7]
```diff
@@ -652,28 +652,28 @@ def test_get_numeric_data(self):
...
         tm.assert_almost_equal(
-            mgr.iget(mgr.items.get_loc("float")).internal_values(),
-            numeric.iget(numeric.items.get_loc("float")).internal_values(),
+            mgr.iget(mgr_idx).internal_values(),
+            numeric.iget(num_idx).internal_values(),
         )
 
         # Check sharing
-        numeric.iset(numeric.items.get_loc("float"), np.array([100.0, 200.0, 300.0]))
+        numeric.iget(num_idx).internal_values()[:] = [100.0, 200.0, 300.0]
```
Comment (jreback) [inline pandas/tests/internals/test_internals.py:666]: i would split the inplace to a separate test

File: pandas/tests/internals/test_internals.py [hunk 8]
```diff
@@ -692,18 +692,20 @@ def test_get_bool_data(self):
...
             mgr.iget(mgr.items.get_loc("bool")).internal_values(),
-            np.array([True, False, True]),
+            np.array([True, True, True]),
         )
 
-        # Check sharing
+        # Check non-sharing
         bools2 = mgr.get_bool_data(copy=True)
-        bools2.iset(0, np.array([False, True, False]))
+        bools2.blocks[0].values[:] = [[False, True, False]]
```
Comment (jreback) [inline pandas/tests/internals/test_internals.py:705]: can you split the non-sharing to a new test

**Review by jreback — CHANGES_REQUESTED** [review]
i actually hadn't looked at this before. on board. still need to look again. did @jorisvandenbossche or @TomAugspurger have objections?

File: pandas/tests/frame/methods/test_rename.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/tests/frame/methods/test_rename.py:165]: the test below this has the copy=True test, but then also has some other stuff.  will look at splitting that test and then parametrizing in a separate pass

**Comment by jorisvandenbossche** [comment]
I didn't yet look back into it, but as far as I know, the discussion I mentioned above (https://github.com/pandas-dev/pandas/pull/35417#issuecomment-681997130) is not yet resolved

**Comment by jbrockmendel** [comment]
@jorisvandenbossche the discussion in #35417 is unresolved only inasmuch as you are OK with the inconsistent behavior bug

Tom's [last](https://github.com/pandas-dev/pandas/issues/33457#issuecomment-664549215) comments was to fix this in 1.2 and jeff's [last](https://github.com/pandas-dev/pandas/issues/33457#issuecomment-660172070) expressed opinion was also to fix this.

File: pandas/tests/internals/test_internals.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/tests/internals/test_internals.py:724]: updated  + green

**Comment by jorisvandenbossche** [comment]
> Tom's last comments was to fix this in 1.2 and jeff's last expressed opinion was also to fix this.

And it's only after those comments that I argued why I don't think we should change this for 1.2, and since then neither Tom or Jeff reacted again.

**Comment by jreback** [comment]
ok moving these off 1.2. IF we made any change like this it cannot be at the last minute.

I am now confused over what we want. What i do know is we DO not want any special cases. I have to read the issue again, but it seems that @jorisvandenbossche is advocating for just that.

**Comment by jorisvandenbossche** [comment]
We have lots of special cases, and I certainly *also* want to get rid of those. But I think the discussion is more on *how* we should get rid of them (can we just "break" those cases or is there another way, do we want to keep such a breaking change for 2.0 or not, ...)

File: pandas/core/indexing.py [hunk 9]
```diff
@@ -675,8 +676,22 @@ def _ensure_listlike_indexer(self, key, axis=None, value=None):
             # GH#38148
             keys = self.obj.columns.union(key, sort=False)
 
+            # Try to get the right dtype when we do this reindex.
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:679]: @phofl any ideas on how to do this more effectively?  in particular, to get the 32 bit builds passing

Comment (phofl) [reply pandas/core/indexing.py:679]: Concerning the 32bit builds: I think we should only infer the fv if value has a dtype or is a scalar (think is the important part here, not quite sure what to expect with lists)

File: pandas/tests/frame/indexing/test_indexing.py [hunk 10]
```diff
@@ -809,7 +809,7 @@ def test_fancy_getitem_slice_mixed(self, float_frame, float_string_frame):
 
         msg = r"\nA value is trying to be set on a copy of a slice from a DataFrame"
         with pytest.raises(com.SettingWithCopyError, match=msg):
-            sliced["C"] = 4.0
+            sliced.loc[:, "C"] = 4.0
```
Comment (jreback) [inline pandas/tests/frame/indexing/test_indexing.py:812]: do the setitems (original) code behave the same (e.g. mutate & raise SettingWtihCopyError)? if so can you tests both indexers

File: pandas/tests/indexing/test_iloc.py [hunk 11]
```diff
@@ -704,7 +704,7 @@ def test_identity_slice_returns_new_object(self):
         assert sliced_df is not original_df
 
         # should be a shallow copy
-        original_df["a"] = [4, 4, 4]
+        original_df.loc[:, "a"] = [4, 4, 4]
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py:707]: same comment as above

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
@jbrockmendel you keep rebasing this, but to repeat what I said above (https://github.com/pandas-dev/pandas/pull/35417#issuecomment-732455747): this still first needs to be discussed further (https://github.com/pandas-dev/pandas/issues/33457). 
Awaiting further discussion, at this point I am -1 on merging this PR.

**Comment by jbrockmendel** [comment]
> you keep rebasing this

Yes, to keep this from drifting.  This seems like a non-sequitur.

> but to repeat what I said above (#35417 (comment)): this still first needs to be discussed further (#33457).

AFAICT the entire discussion consists of the two of us repeating ourselves.  Do you have something new to add?  if not, we're in the same position of "fix inconsistency" vs "keep inconsistency", and de facto it comes down to @jreback.

File: pandas/tests/frame/indexing/test_indexing.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/tests/frame/indexing/test_indexing.py:815]: updated both tests

File: pandas/tests/indexing/test_iloc.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/test_iloc.py:707]: updated both tests

**Comment by jreback** [comment]
@jorisvandenbossche so re-reading this thread. IIUC. @jbrockmendel change here makes EA and non-EA act consistently by effectively *always* copying on assignment (since EA already copy this is making non-EA copy). And since this generally is affecting a single column, we might have to copy the *entire* block (or alternatly split the block which later results in consolidation).

fair? so the issue is essentially performance (as I think we agree on the semantics).

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
i am good with this change. @jorisvandenbossche @TomAugspurger

**Comment by jorisvandenbossche** [comment]
Sorry for the slow reply here, but personally, not much changed regarding my opinion. I don't know how to explain this without being blamed for repeating myself, but I summarized in https://github.com/pandas-dev/pandas/issues/33457#issuecomment-682001989 that this PR has subtle breaking changes + performance impact, with for me not that much benefit with the current BlockManager. So IMO, this is not worth doing in 1.x.

File: pandas/tests/frame/methods/test_rename.py [hunk 12]
```diff
@@ -173,7 +173,7 @@ def test_rename_multiindex(self):
     @td.skip_array_manager_not_yet_implemented  # TODO(ArrayManager) setitem copy/view
     def test_rename_nocopy(self, float_frame):
         renamed = float_frame.rename(columns={"C": "foo"}, copy=False)
-        renamed["foo"] = 1.0
+        renamed["foo"][:] = 1.0
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/methods/test_rename.py:176]: Is this needed? Testing it out with this branch seems to still change the original `float_frame` with `renamed["foo"] = 1.0` as well

Comment (jorisvandenbossche) [reply pandas/tests/frame/methods/test_rename.py:176]: Ah, no, I tested it with a non-renamed columns .. Setting "foo" indeed no longer changes the original frame.

File: pandas/tests/indexing/test_iloc.py [hunk 13]
```diff
@@ -827,7 +827,16 @@ def test_identity_slice_returns_new_object(self, using_array_manager):
...
+        if not using_array_manager:
+            assert (sliced_df["a"] == 4).all()
+        else:
+            # FIXME: what is the expected/desired behavior here?  test it!
+            pass
+
+        # GH#35417 but setting with setitem creates a new array, so
+        #  sliced_df is not changed
+        original_df["a"] = [5, 5, 5]
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/test_iloc.py:839]: This doesn't raise a SettingWithCopy warning? Above a similar sliced case seems to assert a warning: https://github.com/pandas-dev/pandas/pull/35417/files#diff-f3753d3792b1cdb7c319380f7d1f58636ad0f9bbce037eb6920a6f9830412e9cR559

File: pandas/tests/indexing/test_iloc.py [hunk 14]
```diff
@@ -856,11 +865,21 @@ def test_series_indexing_zerodim_np_array(self):
...
         # Mixed dtype ensures we go through take_split_path in setitem_with_indexer
+        if using_array_manager:
+            mark = pytest.mark.xfail(
+                reason="https://github.com/pandas-dev/pandas/issues/33457"
+            )
+            request.node.add_marker(mark)
+
         cat = Categorical(["A", "B", "C"])
-        df = DataFrame({1: cat, 2: [1, 2, 3]})
+        df = DataFrame({1: cat, 2: [1, 2, 3]}, copy=False)
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/test_iloc.py:880]: Is this change related to this PR? (as this seems a test about iloc, not setitem) Or is it this `copy=False` addition that fixes it?

Comment (jbrockmendel) [reply pandas/tests/indexing/test_iloc.py:880]: good catch, will split this off to a different PR

File: pandas/tests/indexing/test_iloc.py (hunk 13 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/test_iloc.py:839]: `original_df["a"] = [5, 5, 5]` here is setting a new array, not overwriting an existing one, whereas the .loc settings in the linked tests are writing _into_ existing arrays

**Comment by jbrockmendel** [comment]
> but I summarized in #33457 (comment) that this PR has subtle breaking changes + performance impact, with for me not that much benefit with the current BlockManager. So IMO, this is not worth doing in 1.x.

One man's "subtle breaking changes" are another man's "fixing clearly incorrect and inconsistent behavior".  Potential performance impact can be addressed.  Decisions about fixing bugs should not depend on implementation details (i.e BlockManager/ArrayManager).

@jreback discussion-wise we are in roughly the same place we were [three months ago](https://github.com/pandas-dev/pandas/pull/35417#issuecomment-752776400).

**Comment by jreback** [comment]
> Sorry for the slow reply here, but personally, not much changed regarding my opinion. I don't know how to explain this without being blamed for repeating myself, but I summarized in [#33457 (comment)](https://github.com/pandas-dev/pandas/issues/33457#issuecomment-682001989) that this PR has subtle breaking changes + performance impact, with for me not that much benefit with the current BlockManager. So IMO, this is not worth doing in 1.x.

@jorisvandenbossche can you clarify. I believe the getitem (copies) are now removed (@jbrockmendel correct me if i am wrong). So this *only* affects setitem.

If something creates a 'subtle' change then please show a clear example. This is a fully tested and announced. We HAVE to move forward with things. Waiting until 2.x is really not constructive for small things. which IMHO this PR does.

**Comment by jorisvandenbossche** [comment]
> I believe the getitem (copies) are now removed. So this only affects setitem.

Yeah, this PR is only about setitem, not getitem

> If something creates a 'subtle' change then please show a clear example. This is a fully tested and announced.

It's fully tested, but there are several test cases that are changed to follow the changed behaviour (and the whatsnew announcement also gives an example). Those all are examples of subtle changes.

One concrete example:

```python
>>> df = pd.DataFrame(np.random.randn(5,3), columns=['a', 'b', 'c'])
>>> sub = df[:]
>>> sub['b'] = 0.0
>>> df
          a    b         c
0  0.903960  0.0 -1.158076
1  1.067373  0.0  0.025295
2  0.355159  0.0 -0.225259
3  1.904736  0.0  0.125681
4  0.378582  0.0 -0.886312
```

The above is with released pandas. With this PR, the parent `df` no longer gets updated.

> One man's "subtle breaking changes" are another man's "fixing clearly incorrect and inconsistent behavior". 

Certainly, and this always a grey area (especially given the many inconsistencies and absence of clear rules). But, historically speaking, you could also say that it the non-consolidatable blocks / EAs that created the "inconsistent", as in a 2D block world, I think it makes a lot of sense to as much as possible preserve the 2D block structure (i.e. in case the types match). Not that it matters much for the current situation, though, but just to point out this is not necessarily "clearly incorrect". 

> Potential performance impact can be addressed. 

Can it be addressed in this PR? Or how would you address it?

> Decisions about fixing bugs should not depend on implementation details (i.e BlockManager/ArrayManager).

For fixing bugs, yes, I agree. But if see this as an API change (that's of course the discussion.. ;)), then for me that can certainly depend on the implementation, since the ArrayManager is exactly a path to introduce API changes.

**Comment by jbrockmendel** [comment]
>> Potential performance impact can be addressed.

> Can it be addressed in this PR? Or how would you address it?

Let's first make sure we're on the same page about what case you're concerned about perf-wise.  IIUC the issue is that when we do `sub["B"] = 0.0` [here](https://github.com/pandas-dev/pandas/pull/35417#issuecomment-823557153), that 

a) may trigger an `np.delete` call in `Block.delete` (making a costly copy) and
b) the resulting DataFrame is no longer consolidated, so subsequent operations that do consolidate will make more copies.

Does that summarize the perf concern accurately?

For a), my thought is that Block.delete should not make a copy and not alter the block in place, but instead should return a list of blocks each containing a view on a subset of columns.

For b), my preference is to consolidate_inplace less (ideally never).  We already do this much less than we did when this PR was opened.  (FWIW the sticking point I've found is DataFrame.values, where not-consolidating would be terrible for perf if .values were being called repeatedly)

**Comment by simonjayhawkins** [comment]
This PR is milestoned 1.3. @jbrockmendel perf concerns outstanding? will you have time to address?

**Comment by jbrockmendel** [comment]
@jorisvandenbossche can you confirm we're on the [same page](https://github.com/pandas-dev/pandas/pull/35417#issuecomment-827945756) about what the perf concerns are?

**Comment by simonjayhawkins** [comment]
@jbrockmendel if the only issue is a perf issue. can we merge and then update later without changing behavior, or would a behavior change be needed?

**Comment by simonjayhawkins** [comment]
and maybe merge master to be sure.

**Comment by jbrockmendel** [comment]
> @jbrockmendel if the only issue is a perf issue. can we merge and then update later without changing behavior, or would a behavior change be needed?

Assuming I understand the perf concerns (@jorisvandenbossche can you confirm, see a few comments back), the fix i have in mind would change some copy/view semantics

**Comment by jreback** [comment]
@jorisvandenbossche ok here?

**Comment by jorisvandenbossche** [comment]
I am not comfortable with merging this right before the RC. 

(and sorry for my slow response here (I have been quite absent the last weeks for several reasons), which is part of why it's delayed until right before the RC ...)

---

That said, in addition I am also not yet convinced that this PR is worth the 1) subtle breaking changes and 2) performance implications. 

> Does that summarize the perf concern accurately?

I think so, yes. The main point is that a setitem operation, which before this PR could be fully inplace, can now trigger a copy of the (almost) full DataFrame (triggered directly or a potential delayed until a next consolidation). This additional data copy has an inherent cost. And I don't think it can be avoided while preserving the actual original intent of this PR (i.e. not overriding existing data).

But it's not only the performance implications, it are also the subtle API changes. And of course it's always a trade-off. In many case a slight performance degradation and some behaviour changes can be worth it for the benefit of the change. I am just not convinced that the benefits of this PR are big enough.

**Comment by jbrockmendel** [comment]
So we're roughly where we've always been: I think fixing this inconsistency is a bugfix, Joris thinks its an API change.

**Comment by simonjayhawkins** [comment]
> So we're roughly where we've always been: I think fixing this inconsistency is a bugfix, Joris thinks its an API change.

there's a dev meeting tomorrow where this can be discussed.

File: pandas/core/indexing.py [hunk 15]
```diff
@@ -693,8 +694,23 @@ def _ensure_listlike_indexer(self, key, axis=None, value=None):
...
```
Comment (jreback) [inline pandas/core/indexing.py:697]: doesn't infer_dtype_from do this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 15 above)
Comment (jbrockmendel) [reply pandas/core/indexing.py:697]: that infers a dtype, but we need to pass a scalar fill_value to reindex_axis

Comment (jbrockmendel) [reply pandas/core/indexing.py]: removed this edit as no longer necessary.  i think at this point all the controversial bits are gone and we're down to just the bugfix.

**Comment by jbrockmendel** [comment]
closing in favor of #45352
